doc/user_guide/introduction_to_psykal.rst

```diff
@@ -0,0 +1,408 @@
+           /)
+      type(func_type) :: meta_funcs(2) = (/               &
+           func_type(W3,   GH_BASIS),                     &
+           func_type(Wchi, GH_DIFF_BASIS)                 &
+           /)
+      integer :: gh_shape = GH_QUADRATURE_XYoZ
+      integer :: operates_on = CELL_COLUMN
+    contains
+      procedure, nopass :: solver_w3_code
+    end type
+  
+  contains
+  
+    subroutine solver_w3_code(nlayers,                                 &
+                              x, rhs,                                  &
+                              chi_1, chi_2, chi_3, ascalar,            &
+                              ndf_w3, undf_w3, map_w3, w3_basis,       &
+                              ndf_w0, undf_w0, map_w0, w0_diff_basis,  &
+                              nqp_h, nqp_v, wqp_h, wqp_v)
+      ...
+    end subroutine solver_w3_code
+  
+  end module w3_solver_kernel_mod
+
+
+Note that the executor method can also be declared as a module procedure
+interface to provide alternative implementations (e.g. different precisions)
+of the kernel code. These are selected as appropriate by the Fortran compiler,
+depending on the precision of the fields being passed to them::
+
+    type, extends(kernel_type) :: integrate_one_kernel 
+      ... 
+      type(...) :: meta_args(...) = (/ ... /) 
+      ... 
+      integer :: operates_on = ... 
+      ... 
+    end type integrate_one_kernel 
+
+    interface ...
+      module procedure ... 
+    end interface   
+
+
+.. _built-ins:
+
+Built-ins
+---------
+
+Built-ins are operations which can be specified within an invoke call in
+the algorithm layer but do not require an associated kernel implementation
+because they are provided by the infrastructure.
+
+These are useful for commonly used operations, as they reduce the amount of
+code that the PSyKAl project has to maintain. In addition, they offer
+potential performance advantages as their implementation can completely
+change for different architectures and the PSy layer is free to implement
+these operations in whatever way it chooses.
+
+The list of supported Built-ins is API-specific and therefore these are
+described under the documentation of each API. In general, PSyclone will
+need to know the types of the arguments being passed to any Built-ins.
+Each API provides a Fortran file that contains the metadata for all
+Built-in operations supported for that API.
+
+.. note:: When a particular Built-in is used, the name of this
+          Built-in should not be used for anything else within the
+          same scope. For example, it is not valid to make use of a
+          Built-in called ``setval_c`` and for its parent subroutine
+          to also be called ``setval_c``. In this case PSyclone will
+          raise an exception.
+
+Example
++++++++
+
+In the following algorithm-layer example from the LFRic API, the invoke call includes a call
+to two Built-ins (``setval_c`` and ``X_divideby_Y``) and a user-supplied kernel
```

Comment:
As Oakley is currenly implementing support for user-supplied kernels that iterate over dofs, we should change this a bit. Perhaps "user-supplied kernel that operates on cell columns..."?

Reply:
Done